### ISSUE TITLE
Auth: cyclic reference error if Api depends on the Cognito User Pool

### ISSUE THREAD

**Issue by fwang** [serverless-stack/sst#776]
### Issue
```js
const api = new Api(this, "MyApi", {
  defaultFunctionProps: {
    environment: {
      USER_POOL_ID: auth.cognitoUserPool.userPoolId
    }
  }
});
auth.addPermissionsToAuthRole([api]);
```
This causes cyclic reference.

### Workaround
```js
const policy = new iam.Policy(this, "AuthPolicy", {
  statements: [
    new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["execute-api:Invoke"],
      resources: [`arn:aws:execute-api:${scope.region}:${scope.account}:${api.httpApiId}/*`]
    })
  ],
});
policy.attachToRole(auth.iamAuthRole);
```

### Explaination

In the current way `auth` depends on `api`, because `api.httpApiId` is added to `auth`'s IAM role.
```js
auth.addPermissionsToAuthRole([api]);
```

In the workaround way `api` does **NOT** depend on `auth`, instead a new resources `policy` is created and it depends on both `auth` and `api`.
```js
policy.attachToRole(auth.iamAuthRole);
```

### Solution
- Can we detect that the Api stack depends on the Auth stack? If so, we use the workaround way behind the scene.
- Or, we always go with the workaround.

---
Requests:
- George Evans https://serverless-stack.slack.com/archives/C01JG3B20RY/p1630511467176300
- gio https://serverless-stack.slack.com/archives/C01JG3B20RY/p1641562404068600

**Comment by t-indie** [comment]
+1 I also have the same issue

**Comment by ari-becker** [comment]
Fixed (for me) by https://github.com/serverless-stack/sst/pull/1871 and released in v1.3.0, recommend closing

**Comment by fwang** [comment]
Thanks @ari-becker! Indeed!

Leaving a link here to [v1.3.0](https://github.com/serverless-stack/sst/releases/tag/v1.3.0). You can read more about the issue and the fix in the release note.
